utils/src/test/java/com/cloud/utils/testcase/NioTest.java

```diff
@@ -19,146 +19,198 @@
+            }
             try {
-                s_logger.debug(_completedCount + "/" + _testCount + " tests done. Waiting for completion");
+                LOGGER.debug(completedTestCount + "/" + totalTestCount + " tests done. Waiting for completion");
                 Thread.sleep(1000);
             } catch (final InterruptedException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
+                Assert.fail(e.getMessage());
             }
         }
-        stopClient();
-        stopServer();
+        LOGGER.debug(completedTestCount + "/" + totalTestCount + " tests done.");
     }
 
-    protected void stopClient() {
-        _client.stop();
-        s_logger.info("Client stopped.");
+    protected void doServerProcess(final byte[] data) {
+        oneMoreTestDone();
+        Assert.assertArrayEquals(testBytes, data);
+        LOGGER.info("Verify data received by server done.");
     }
 
-    protected void stopServer() {
-        _server.stop();
-        s_logger.info("Server stopped.");
+    public byte[] getTestBytes() {
+        return testBytes;
     }
 
-    protected void setClientLink(final Link link) {
-        _clientLink = link;
+    public class ThreadedNioClient implements Runnable {
+        final private NioClient client;
+        ThreadedNioClient(final NioClient client) {
+            this.client = client;
+        }
+
+        @Override
+        public void run() {
+            try {
+                client.start();
+            } catch (NioConnectionException e) {
+                Assert.fail(e.getMessage());
+            }
+        }
     }
 
-    Random randomGenerator = new Random();
+    public class NioMaliciousClient extends NioClient {
 
-    byte[] _testBytes;
+        public NioMaliciousClient(String name, String host, int port, int workers, HandlerFactory factory) {
+            super(name, host, port, workers, factory);
+        }
 
-    public void testConnection() {
-        _testBytes = new byte[1000000];
-        randomGenerator.nextBytes(_testBytes);
-        try {
-            getOneMoreTest();
-            _clientLink.send(_testBytes);
-            s_logger.info("Client: Data sent");
-            getOneMoreTest();
-            _clientLink.send(_testBytes);
-            s_logger.info("Client: Data sent");
-        } catch (final ClosedChannelException e) {
-            // TODO Auto-generated catch block
-            e.printStackTrace();
+        @Override
+        protected void init() throws IOException {
+            _selector = Selector.open();
+            try {
+                _clientConnection = SocketChannel.open();
+                LOGGER.info("Connecting to " + _host + ":" + _port);
+                final InetSocketAddress peerAddr = new InetSocketAddress(_host, _port);
+                _clientConnection.connect(peerAddr);
+                // Hang in there don't do anything
+                Thread.sleep(3600000);
```

Comment:
(facepalm) can you read the test, understand what/how/why, thanks.

The whole idea of this malicious client is to simply connect to the server but no participate in SSL handshake and do any IO. The test fails if it does not finish in 600s.

Code review is line-visioned when diff is viewed in unified mode, I recommend the split mode view to see the bigger picture.